### PR TITLE
Add OpenTelemetry tracing instrumentation to endpoints handlers

### PR DESCRIPTION
**Pull request by rannyeli**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✖                                                                       |

Further  information:

Added deeper tracing instrumentation to `/layers` and `/:jobId/:taskId/completed` routes.
image

Next PRs:
Will add traces to other services, and try to combine traces between this service and the others.

### CODE REVIEW

File: src/common/tracing.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import { Tracing } from '@map-colonies/telemetry';
 import { ExpressInstrumentation } from '@opentelemetry/instrumentation-express';
 import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
+
```
Comment (asafMasa) [inline src/common/tracing.ts]: remove extra line

File: src/layers/models/layersManager.ts [hunk 2]
```diff
@@ -130,25 +146,33 @@ export class LayersManager {
           throw new ConflictError(message);
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts]: Don't we need to add here "handleSpanOnError"

File: src/layers/models/layersManager.ts [hunk 3]
```diff
@@ -163,7 +187,7 @@ export class LayersManager {
 
         // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
       } else if (jobType === JobAction.UPDATE) {
-        const record = await this.catalog.findRecord(productId, undefined, productType);
+        const record = await asyncCallInSpan(async () => this.catalog.findRecord(productId, undefined, productType), this.tracer, 'findRecord');
```
Comment (asafMasa) [inline src/layers/models/layersManager.ts]: Don't we need to add "handleSpanOnError"  on "if (!existsInMapProxy) .... throw new BadRequestError(message);" several lines ahead?

**Review by asafMasa — CHANGES_REQUESTED** [review]
Looking Good, Thanks! 😎 
Note: 
* Please fix comments and tests 
* Check that "enabled" false is working
  * Why do you create a span if in false?

File: src/layers/models/layersManager.ts (hunk 2 above)
Comment (rannyeli) [reply src/layers/models/layersManager.ts]: This code located within a try-catch block, so I included the `handleSpanOnError` within the catch scope. As any errors are thrown, they will pass through `handleSpanOnError` while the error itself is caught.

File: src/layers/models/layersManager.ts (hunk 3 above)
Comment (rannyeli) [reply src/layers/models/layersManager.ts]: and the same here. this code is under the same try-catch block

File: helm/values.yaml [hunk 4]
```diff
@@ -108,7 +108,7 @@ env:
     mountPath: /usr/app/config
     fileName: linkTemplates.template
   tracing:
-    enabled: false
+    enabled: true
```
Comment (asafMasa) [inline helm/values.yaml]: default should be false

File: src/serviceClients/catalogClient.ts [hunk 5]
```diff
@@ -22,42 +28,43 @@ export class CatalogClient extends HttpClient {
...
+      return undefined;
+    }
 
-    return res;
+    // Return metadata
+    return res[0];
   }
 
-  public async findRecord(productId: string, productVersion?: string, productType?: string): Promise<IFindResponseRecord | undefined> {
+  public async exists(productId: string, productVersion?: string, productType?: string): Promise<boolean> {
```
Comment (asafMasa) [inline src/serviceClients/catalogClient.ts:53]: add decorator

File: src/serviceClients/catalogClient.ts [hunk 6]
```diff
@@ -22,42 +28,43 @@ export class CatalogClient extends HttpClient {
...
-    // Check if product exists with given version
-    if (res.length == 0) {
-      return undefined;
-    }
+    return res.length > 0;
+  }
 
-    // Return metadata
-    return res[0];
+  public async existsByRecordId(recordId: string): Promise<boolean> {
```
Comment (asafMasa) [inline src/serviceClients/catalogClient.ts:66]: add decorator

File: src/serviceClients/catalogClient.ts [hunk 7]
```diff
@@ -22,42 +28,43 @@ export class CatalogClient extends HttpClient {
...
+    const res = await this.get<boolean>(`/records/exists/${recordId}`);
+
+    return res;
   }
 
   public async publish(record: IRasterCatalogUpsertRequestBody): Promise<string> {
```
Comment (asafMasa) [inline src/serviceClients/catalogClient.ts:73]: add decorator for all external API's (update, publish, getLayerVersions, etc.)

File: src/utils/zoomToResolution.ts [hunk 8]
```diff
@@ -8,11 +10,12 @@ import { IConfig } from '../common/interfaces';
 export class ZoomLevelCalculator {
   private readonly zoomRanges: ITaskZoomRange[];
 
-  public constructor(@inject(SERVICES.CONFIG) private readonly config: IConfig) {
+  public constructor(@inject(SERVICES.CONFIG) private readonly config: IConfig, @inject(SERVICES.TRACER) public readonly tracer: Tracer) {
     const batches = config.get<string[]>('tiling.zoomGroups');
     this.zoomRanges = this.getZoomRanges(batches);
   }
 
+  @withSpanV4
```
Comment (asafMasa) [inline src/utils/zoomToResolution.ts]: no need for this data

**Review by asafMasa — CHANGES_REQUESTED** [review]
Looking good

**Review by asafMasa — APPROVED** [review]
Thanks!

**Pull request merged**